beacon_node/network/src/sync/sampling.rs

```diff
@@ -0,0 +1,486 @@
+        // Select columns to sample
+        // Create individual request per column
+        // Progress requests
+        // If request fails retry or expand search
+        // If all good return
+        let Some(request) = self.column_requests.get_mut(&column_index) else {
+            warn!(
+                self.log,
+                "Received sampling response for unrequested column index"
+            );
+            return Ok(None);
+        };
+
+        match resp {
+            Ok((mut data_columns, seen_timestamp)) => {
+                debug!(self.log, "Sample download success"; "block_root" => %self.block_root, "column_index" => column_index, "count" => data_columns.len());
+
+                // No need to check data_columns has len > 1, as the SyncNetworkContext ensure that
+                // only requested is returned (or none);
+                if let Some(data_column) = data_columns.pop() {
+                    // Peer has data column, send to verify
+                    let Some(beacon_processor) = cx.beacon_processor_if_enabled() else {
+                        // If processor is not available, error the entire sampling
+                        debug!(self.log, "Dropping sampling"; "block" => %self.block_root, "reason" => "beacon processor unavailable");
+                        return Err(SamplingError::ProcessorUnavailable);
+                    };
+
+                    debug!(self.log, "Sending data_column for verification"; "block" => ?self.block_root, "column_index" => column_index);
+                    if let Err(e) = beacon_processor.send_rpc_data_columns(
+                        self.block_root,
+                        vec![data_column],
+                        seen_timestamp,
+                        SamplingId {
+                            id: self.requester_id,
+                            column_index,
+                        },
+                    ) {
+                        // TODO(das): Beacon processor is overloaded, what should we do?
+                        error!(self.log, "Dropping sampling"; "block" => %self.block_root, "reason" => e.to_string());
+                        return Err(SamplingError::SendFailed("beacon processor send failure"));
+                    }
+                } else {
+                    // Peer does not have the requested data.
+                    // TODO(das) what to do?
+                    debug!(self.log, "Sampling peer claims to not have the data"; "block_root" => %self.block_root, "column_index" => column_index);
+                    request.on_sampling_error()?;
+                }
+            }
+            Err(err) => {
+                debug!(self.log, "Sample download error"; "block_root" => %self.block_root, "column_index" => column_index, "error" => %err);
+
+                // Error downloading, maybe penalize peer and retry again.
+                // TODO(das) with different peer or different peer?
+                request.on_sampling_error()?;
```

Comment:
Do we currently or plan to keep track of the peers that fail to return sample results somewhere and penalise if it's recurring?

Reply:
Scoring in sampling is a complicated topic. There's no clear attribution if the proposer is playing delay games to force donwscoring events or the peers are being actively malicious. For now I would just do nothing and figure it out latter in the interop

Reply:
Sounds good